src/dune_rules/lib.ml

```diff
@@ -1894,11 +1976,38 @@ module DB = struct
 
   let available t name = Resolve_names.available_internal t name
 
-  let get_compile_info t ~allow_overlaps name =
+  let available_library_stanza t ~dir name =
+    let open Memo.O in
+    available t name
+    >>= function
+    | false -> Memo.return false
+    | true ->
+      find t name
+      >>| (function
+       | None -> assert false
+       | Some lib ->
+         (match Local.of_lib lib with
+          | None -> false
+          | Some info -> Path.Build.equal dir (Lib_info.src_dir (Local.info info))))
```

Comment:
This works at the moment because the code assumes two libs with the same name must exist in different folders, but this might not be necessarily true in the future, see related comment: https://github.com/ocaml/dune/pull/9839#issuecomment-1941386688